inst/include/Rcpp/vector/no_init.h

```diff
@@ -35,9 +35,12 @@ class no_init_vector {
         return size;
     }
 
-    template <int RTYPE, template <class> class StoragePolicy >
-    operator Vector<RTYPE, StoragePolicy>() const {
-        return Rf_allocVector(RTYPE, size) ;
+    template <int RTYPE>
+    operator Vector<RTYPE, PreserveStorage>() const {
```

Comment:
Why change the template parameters here? Do we need separate operators for each of the storage policies?

Reply:
I think `no_init()` should always return a properly protected vector, that's why I removed this template parameter. But I'm not sure.

Reply:
The intention is that the caller can indicate whether they want protection or want -- by default, they get a protected vector; if they don't want one then they can request conversion to the `NoPreserveStorage` variant.

IIUC, since this function was previously returning a raw SEXP, the associated constructor would be called on that return value and that constructor would take care of preserving the underlying storage when required.

Reply:
As I recall, the "feature" of `no_init()` was to save the handful of cycles needed to set the memory to zero, but not to alter anything on protection or lack thereof.

Reply:
`no_init()` effectively returns an uninitialized vector (via constructing a temporary and calling `operator Vector<...>()` on it). To be on the very safe side, this uninitialized vector needs to be protected during its lifetime; the protection state of the target is of no relevance. This corner case is perhaps irrelevant.

`Vector::Vector(SEXP)` seems to coerce (in the very generic case, not affecting us, but may be hard to tell for `rchk`). Maybe `rchk` thinks (or sees) that this constructor is involved in `Vector<...> x = no_init(...)`?

Happy to revert to the original implementation (with the second template parameter), but I'll check first if `rchk` eats this.